# 1. Checking data edges

In [1]:
import pandas as pd
df = pd.read_csv('/data/datasets/cleanhouses.csv')

In [2]:
len(df)

1460

In [9]:
df.describe()

Id   MSSubClass  LotFrontage        LotArea  OverallQual  \
count  1460.000000  1460.000000  1201.000000    1460.000000  1460.000000   
mean    730.500000    56.897260    70.049958   10516.828082     6.099315   
std     421.610009    42.300571    24.284752    9981.264932     1.382997   
min       1.000000    20.000000    21.000000    1300.000000     1.000000   
25%     365.750000    20.000000    59.000000    7553.500000     5.000000   
50%     730.500000    50.000000    69.000000    9478.500000     6.000000   
75%    1095.250000    70.000000    80.000000   11601.500000     7.000000   
max    1460.000000   190.000000   313.000000  215245.000000    10.000000   

       OverallCond    YearBuilt  YearRemodAdd   MasVnrArea   BsmtFinSF1  ...  \
count  1460.000000  1460.000000   1460.000000  1452.000000  1460.000000  ...   
mean      5.575342  1971.267808   1984.865753   103.685262   443.639726  ...   
std       1.112799    30.202904     20.645407   181.066207   456.098091  ...   
min       1.000000  1872.000000   1950.000000     0.000000     0.000000  ...   
25%       5.000000  1954.000000   1967.000000     0.000000     0.000000  ...   
50%       5.000000  1973.000000   1994.000000     0.000000   383.500000  ...   
75%       6.000000  2000.000000   2004.000000   166.000000   712.250000  ...   
max       9.000000  2010.000000   2010.000000  1600.000000  5644.000000  ...   

        WoodDeckSF  OpenPorchSF  EnclosedPorch    3SsnPorch  ScreenPorch  \
count  1460.000000  1460.000000    1460.000000  1460.000000  1460.000000   
mean     94.244521    46.660274      21.954110     3.409589    15.060959   
std     125.338794    66.256028      61.119149    29.317331    55.757415   
min       0.000000     0.000000       0.000000     0.000000     0.000000   
25%       0.000000     0.000000       0.000000     0.000000     0.000000   
50%       0.000000    25.000000       0.000000     0.000000     0.000000   
75%     168.000000    68.000000       0.000000     0.000000     0.000000   
max     857.000000   547.000000     552.000000   508.000000   480.000000   

          PoolArea       MiscVal       MoSold       YrSold      SalePrice  
count  1460.000000   1460.000000  1460.000000  1460.000000    1460.000000  
mean      2.758904     43.489041     6.321918  2007.815753  180921.195890  
std      40.177307    496.123024     2.703626     1.328095   79442.502883  
min       0.000000      0.000000     1.000000  2006.000000   34900.000000  
25%       0.000000      0.000000     5.000000  2007.000000  129975.000000  
50%       0.000000      0.000000     6.000000  2008.000000  163000.000000  
75%       0.000000      0.000000     8.000000  2009.000000  214000.000000  
max     738.000000  15500.000000    12.000000  2010.000000  755000.000000  

[8 rows x 38 columns]

In [8]:
pd.DataFrame( {'min':df.min(), 'max':df.max()} )

min      max
Id                   1     1460
MSSubClass          20      190
MSZoning       C (all)       RM
LotFrontage         21      313
LotArea           1300   215245
Street            Grvl     Pave
LotShape           IR1      Reg
LandContour        Bnk      Lvl
Utilities       AllPub   NoSeWa
LotConfig       Corner   Inside
LandSlope          Gtl      Sev
Neighborhood   Blmngtn  Veenker
Condition1      Artery     RRNn
Condition2      Artery     RRNn
BldgType          1Fam   TwnhsE
HouseStyle      1.5Fin     SLvl
OverallQual          1       10
OverallCond          1        9
YearBuilt         1872     2010
YearRemodAdd      1950     2010
RoofStyle         Flat     Shed
RoofMatl       ClyTile  WdShngl
Exterior1st    AsbShng  WdShing
Exterior2nd    AsbShng  Wd Shng
MasVnrArea           0     1600
ExterQual           Ex       TA
ExterCond           Ex       TA
Foundation      BrkTil     Wood
BsmtFinSF1           0     5644
BsmtFinSF2           0     1474
...                ...      ...
1stFlrSF           334     4692
2ndFlrSF             0     2065
LowQualFinSF         0      572
GrLivArea          334     5642
BsmtFullBath         0        3
BsmtHalfBath         0        2
FullBath             0        3
HalfBath             0        2
BedroomAbvGr         0        8
KitchenAbvGr         0        3
KitchenQual         Ex       TA
TotRmsAbvGrd         2       14
Functional        Maj1      Typ
Fireplaces           0        3
GarageYrBlt       1900     2010
GarageCars           0        4
GarageArea           0     1418
PavedDrive           N        Y
WoodDeckSF           0      857
OpenPorchSF          0      547
EnclosedPorch        0      552
3SsnPorch            0      508
ScreenPorch          0      480
PoolArea             0      738
MiscVal              0    15500
MoSold               1       12
YrSold            2006     2010
SaleType           COD       WD
SaleCondition  Abnorml  Partial
SalePrice        34900   755000

[65 rows x 2 columns]

Pandas reads the file as a `Dataframe` object, which resembles a table in a database. The header (first line) provides the column labels, and every row in the file is inserted as a row in the Dataframe. We can view the first view lines with the `head()` method. You can also use `head(n)` to show the first `n` lines.

# the number of rows of a dataframe

If you know how many samples there should be (for instance from the source where you obtained the data) check it.

#### Assignment: Enter code to get the number of rows in the dataframe using the len() function, the correct answer should be 1460.

In [3]:
len(df)

1460

# columns in the dataframe

Similarly, if you known how many features there should be, check it. Also check for key features that you need for the problem you want to study. In this dataset, since the task is to predict the target variable SalePrice, we would require features that we expect to be most useful such as the size of the house.

In [ ]:
df.columns

# the number of columns in a dataframe

In [ ]:
len(df.columns)

### Inspecting the first and last rows

One of the reasons you want to inpect the first and last rows is to make sure all the data was read ok. Sometimes, files are split, meaning you end up with half a record, or someone inserted some commentary on the first or last lines.

In [ ]:
df.head() # first rows

#### Assignment: Enter the code to only inspect the first row, hint you can pass the number of rows to head()

We can view the datatypes and the number of non-null values, using the `info()` method. Null refers to 'no value' or 'unkown value'.

In [ ]:
df.tail() # last 5 rows

### subsetting rows

In [ ]:
# select the first two rows (the upper bound is always exclusive)
df[:2]

In [ ]:
# select the last 2 rows, negative numbers are an index that count back from the end of the dataframe
df[-2:]

#### Assignment: select rows 2 and 3 in one statement

### Some additional inspections
### Subset rows by boolean indexing (logical statements)

In [ ]:
# creates a boolean index that indicates which rows have '04' in ExpMM.
df.MSZoning == 'RL'

In [ ]:
# create a subset dataframe of only rows that contain 'IR1' for the column LotShape
dfrl = df[df.LotShape == 'IR1']
dfrl.head(4)

In [ ]:
# create a subset dataframe of only rows that contain
# 'IR1' for the column LotShape AND '2008' for YrSold
dfrlreg = df[(df.LotShape == 'IR1') | (df.YrSold == 2008)]
dfrlreg.head(4)

In [ ]:
# create a subset dataframe of only rows that contain
# IR1' for the column LotShape OR '2008' for YrSold
dfrlreg = df[(df.LotShape == 'IR1') | (df.YrSold == 2008)]
dfrlreg.head(4)

#### Assignment: Show the sizes of the poolareas of the 7 houses that have a pool area (tip: PoolArea > 0)

### Subsetting columns
More advanced subsetting can be done with df.ix. The first range selects the rows (: selects all). The second range selects columns, either by index number of label name. Somewhat confusing, the upperbounds are inclusive here.

In [ ]:
# More advanced subsetting can be done with df.ix. The first range selects the rows. 
df.ix[:5, 'LotArea':'LotShape']

In [ ]:
# alternatively, give a list of columns.
df[['1stFlrSF', '2ndFlrSF']]